### ISSUE TITLE
New Octane Rule: `no-component-lifecycle-hooks`

### ISSUE THREAD

**Issue by patocallaghan** [ember-cli/eslint-plugin-ember#540]
Pulled from https://github.com/ember-cli/eslint-plugin-ember/issues/533 which proposes a new set of rules for Octane related best practices

Disallow "Classic" Ember Component lifecycle hooks. Prefer using `@ember/render-modifiers` or custom functional modifiers.

List of Component lifecycle hooks pulled from [here](https://guides.emberjs.com/release/components/the-component-lifecycle/#toc_order-of-lifecycle-hooks-called)

```js
// ### Good
export default class MyComponent extends Component {
  init() { ... }
  willDestroy() { ... }
}

export default Component.extend({
  init() { ... }
  willDestroy() { ... }
});

// ### Bad
export default class MyComponent extends Component {
  didDestroyElement() {}
  didInsertElement() {}
  didReceiveAttrs() {}
  didRender() {}
  didUpdate() {}
  didUpdateAttrs() {}
  willClearRender() {}
  willDestroyElement() {}
  willRender() {}
}

export default Component.extend({
  didDestroyElement() {},
  didInsertElement() {},
  didReceiveAttrs() {},
  didRender() {},
  didUpdate() {},
  didUpdateAttrs() {},
  willClearRender() {},
  willDestroyElement() {},
  willRender() {},
});
```

If anyone would like to implement this please comment that you'll take it and work away.

**Comment by patocallaghan** [comment]
@rwjblue @Turbo87 could you sanity check the good/bad examples please?

Do we also need to check that these occur in Component class or is it good enough to just look for function names?

**Comment by scalvert** [comment]
This seems reasonable. It's pretty crazy how much the API is narrowing...

**Comment by jbandura** [comment]
I’d like to tackle this 🙂

**Comment by patocallaghan** [comment]
@jbandura great 🙌 

@scalvert @rwjblue just thinking, in Glimmer components should we also be linting against the `init` hook, i.e. use `constructor` instead?

**Comment by alexlafroscia** [comment]
Do Glimmer components even have an `init` hook that would be called?

The run-time behavior not working might be enough -- trying to lint against using a hook that doesn't exist at all feels like starting to push into the realm of type-checking rather than linting.

**Comment by jbandura** [comment]
Quick question regarding how smart we want the rule to be. In the proposed example:

```js
// ### Bad
export default class MyComponent extends Component {
  didDestroyElement() {}
}
```
This could, in theory, be either a `Component` imported from `@glimmer/component` or from `@ember/component`. In the latter case, we do not necessarily want to complain about the usage of the lifecycle hooks, as they're still present in the "classic" ember components, right? I guess we should also take that into consideration in the rule? 

One example of a situation where both types of components would co-exist would be migrating an old application to Octane. We would then have both classic components and octane components living side by side. Does this make sense, or did I miss something?

**Comment by Turbo87** [comment]
@jbandura yes, that was the intention. the new way of doing the same thing via modifiers works for ember components too, so IMHO we should warn about the life cycle hooks for those too
